nixos/nfs: Allow Kerberized NFS

kwohlfahrt opened the pull request:
###### Motivation for this change

Currently, NFS mounts secured with Kerberos do not work (see #72722).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @tfc - regarding the Perl -> Python test port
cc @edolstra as maintainer of the previous nfs test

kwohlfahrt committed:
nixos/nfs: convert test to python

rnhmjoj reviewed (changes requested):
You have to update `nixos/tests/all-tests.nix` to build the `nfs/` directory, instead of `nfs.nix`.

kwohlfahrt commented on nixos/tests/nfs/kerberos.nix:
```diff
@@ -0,0 +1,109 @@
+      client.fail("echo bla >> /data/foo")
+
+      client.succeed("echo root_pw | kinit root", "echo bla >> /data/foo")
+      server.succeed("test -e /data/foo")
```
`test -e` is what the plain NFS test uses, I assume because NFS should be synchronous here (i.e. by the time the file is visible on the client, it has been created on the server).

kwohlfahrt commented on nixos/tests/nfs/kerberos.nix:
```diff
@@ -0,0 +1,109 @@
+
+      # start client
+      client.succeed("systemctl start rpc-gssd.service")
+      client.wait_for_unit(f"rpc-gssd.service")
```
Fixed.

kwohlfahrt commented on nixos/tests/nfs/kerberos.nix:
```diff
@@ -0,0 +1,109 @@
+      client.succeed("systemctl start rpc-gssd.service")
+      client.wait_for_unit(f"rpc-gssd.service")
+
+      # add principals to server keytab
```
Fixed.

rnhmjoj commented on nixos/tests/all-tests.nix:
```diff
@@ -187,8 +187,9 @@ in
-  nfs4 = handleTest ./nfs.nix { version = 4; };
+  # TODO: Test nfsv3 + Kerberos
+  nfs3 = handleTest ./nfs/simple.nix { version = 3; };
+  nfs4 = handleTest ./nfs { version = 4 };
```
You're missing a semicolon here.

kwohlfahrt replied:
Fixed.

kwohlfahrt commented:
As noted in the issue, there are two remaining problems:

1. file uid/gids are all messed up. This can be fixed by running nfs-idmapd on the client, but [docs](https://dyn.manpages.debian.org/unstable/nfs-common/idmapd.8.en.html) suggest that this is no longer necessary as it is replaced by the nfsidmap binary. Indeed, the service is masked if `services.nfs.server.enable = false`. I'm going to try on a few different distros to see what they do.
2. It is not possible to write to files as root, even if the corresponding kerberos principal exists. This may be intended behavior.

tfc commented on nixos/tests/nfs/simple.nix:
```diff
@@ -41,50 +41,55 @@ in
+      server.succeed("test -e /data/bar")
 
       # Test whether restarting ‘nfs-server’ works correctly.
-      $server->succeed("systemctl restart nfs-server");
```
The comment could be substituted with a `subtest`, then even the log will contain this information.

kwohlfahrt replied:
Fixed.

kwohlfahrt commented:
I appear to have CI errors, but the message in details is not very clear to me: `attribute 'x86_64-linux' missing, at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/eval-0-gleber.ewr1.nix.ci/nixos/release-combined.nix:50:43`. How do I fix this?

kwohlfahrt commented:
This PR has expanded in scope a bit - there is now a kernel patch to fix the hard-coded ID `request-key` path. This touches a few other issues, namely #68106 and #34638, though those still need a bit of module work to ensure `keyutils` is present for example.

tfc commented on nixos/tests/nfs/kerberos.nix:
```diff
@@ -0,0 +1,134 @@
+      server.succeed("mkdir -p /data/alice")
+      server.succeed("chown alice:users /data/alice")
+
+      # set up kerberos database
```
the comments here that tell what happens on a higher level could all be `subtest`s, so this information even appears in the log.

tfc replied:
Oh sorry i already stated this in another comment in the same PR and didn't remember.

kwohlfahrt replied:
I've changed most of the other ones. Here, this is really set-up code that is not related to the functionality under test. Is `subtest` intended for  this use as well?

rnhmjoj commented:
> This PR has expanded in scope a bit - there is now a kernel patch to fix the hard-coded ID request-key path. This touches a few other issues, namely #68106 and #34638, though those still need a bit of module work to ensure keyutils is present for example.

I had no idea this could get this complicated. Thanks for all the work you've done, keep it up!

rnhmjoj commented:
> I appear to have CI errors, but the message in details is not very clear to me: attribute 'x86_64-linux' missing, at /var/lib/ofborg/checkout/repo/38dca4e3aa6bca43ea96d2fcc04e8229/mr-est/eval-0-gleber.ewr1.nix.ci/nixos/release-combined.nix:50:43. How do I fix this?

I can't try right now because I don't have nearly ram to run an evaluation but you could try to rebase your branch with current master. Maybe at the point you did a checkout master itself was broken.

ofBorg runs something like this

    $ nix-instantiate nixos/release-combined.nix -A tested

kwohlfahrt commented:
I figured out the build failure - I copied the test structure from the kerberos tests, which I wrote last year. It turns out they were never enabled in `release-combined.nix`, because I didn't know this was a necessary step, and the way I'd been calling them doesn't work.

kwohlfahrt committed:
nfs-utils: build svcgssd

Necessary for nfs server with kerberos auth

kwohlfahrt committed:
nixos/nfs: test nfs with kerberos authentication

kwohlfahrt committed:
linux: patch request-key binary path

This is necessary for id mapping to work with NFS + Kerberos, and also
touches #68106 and 634638.

kwohlfahrt commented:
All of the tests here are OK now. Remaining question are:

- How should `/etc/request-key.conf` be generated? The issues linked above require different lines in this file.
- should `keyutils` be part of the default `systemPackages`?

rnhmjoj commented:
> should keyutils be part of the default systemPackages?

If it's only needed for kerberos I would say to install it behind a guard like `krb5.enable`.

kwohlfahrt commented:
@rnhmjoj - no, its needed for id-mapping in general. This is not turned on by default for `sec=sys` (i.e. not kerberos), but such a configuration is possible, see [here](https://unix.stackexchange.com/questions/438939/how-to-get-nfsv4-idmap-working-with-sec-sys). I assume this is the reason for #68106 as it doesn't mention kerberos anywhere.

It is also needed for CIFS, at least with kerberos, see #34638. I'm not familiar enough with it to know if this can also be necessary without kerberos - the [man page](https://linux.die.net/man/8/cifs.idmap) doesn't mention it so it seems likely.

We can test for the presence of a NFS/CIFS filesystem before installing it, which seems sensible.

kwohlfahrt committed:
keyutils: patch to allow symlinks in config dir

kwohlfahrt committed:
nfs: set up request-key for id mapping

A patch is necessary upstream to support multiple configs via symlinks
in /etc/request-key.d

Once that is done, we can add support for CIFS as well

kwohlfahrt commented:
With the patch to `keyutils`, this causes a mass-rebuild so I have changed the base to `staging`.

Mic92 commented on nixos/tests/nfs/kerberos.nix:
```diff
@@ -0,0 +1,133 @@
+          "systemctl restart kadmind.service kdc.service",
+      )
+      for unit in ["kadmind", "kdc"]:
+          server.wait_for_unit(f"{unit}.service")
```
nitpick: same lines of code / easier to read:

```suggestion
    server.wait_for_unit("kadmind.service")
    server.wait_for_unit("kdc.service")
```

kwohlfahrt replied:
Nits picked. Anything left blocking merging?

Mic92 commented:
@GrahamcOfBorg test nfs

Mic92 commented:
@GrahamcOfBorg test nfs.simple nfs.kerberos

Mic92 commented:
@GrahamcOfBorg test nfs3.simple nfs4.simple nfs4.kerberos

Mic92 committed:
nfs-utils: link nfs tests
